File: app/src/main/java/com/nikola/jakshic/dagger/view/adapter/PeerAdapter.java
package com.nikola.jakshic.dagger.view.adapter;

import android.arch.paging.PagedListAdapter;
import android.content.Context;
import android.content.Intent;
import android.support.annotation.NonNull;
import android.support.v7.util.DiffUtil;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import com.nikola.jakshic.dagger.databinding.ItemPeerBinding;
import com.nikola.jakshic.dagger.inspector.PeerInspector;
import com.nikola.jakshic.dagger.model.Peer;
import com.nikola.jakshic.dagger.view.activity.PlayerActivity;

public class PeerAdapter extends PagedListAdapter<Peer, PeerAdapter.PeerViewHolder> {

    private Context context;

    public PeerAdapter(Context context, @NonNull DiffUtil.ItemCallback<Peer> diffCallback) {
        super(diffCallback);
        this.context = context;
    }

    @NonNull
    @Override
    public PeerViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        ItemPeerBinding binding = ItemPeerBinding.inflate(
                LayoutInflater.from(parent.getContext()),
                parent,
                false);
        return new PeerViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(@NonNull PeerViewHolder holder, int position) {
        Peer item = getItem(position);
        holder.binding.setInspector(new PeerInspector(context, item));
        // TODO move into viewHolder
        holder.binding.getRoot().setOnClickListener(v -> {
            Intent intent = new Intent(context, PlayerActivity.class);
            intent.putExtra("player-personaname", item.getPersonaname());
            intent.putExtra("player-account-id", item.getPeerId());
            intent.putExtra("player-avatar-full", item.getAvatarfull());
            context.startActivity(intent);
        });
    }

    public class PeerViewHolder extends RecyclerView.ViewHolder {

        private ItemPeerBinding binding;

        public PeerViewHolder(ItemPeerBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }
    }
}

File: app/src/main/java/com/nikola/jakshic/dagger/profile/peers/PeerAdapter.kt
package com.nikola.jakshic.dagger.profile.peers

import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import coil.load
import coil.transform.CircleCropTransformation
import com.nikola.jakshic.dagger.R
import com.nikola.jakshic.dagger.common.inflate
import com.nikola.jakshic.dagger.databinding.ItemPeerBinding

class PeerAdapter(
    private val listener: (Long) -> Unit
) : ListAdapter<PeerUI, PeerAdapter.ViewHolder>(DIFF_CALLBACK) {
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        return ViewHolder(parent.inflate(R.layout.item_peer))
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.bind(getItem(position))
    }

    inner class ViewHolder(view: View) : RecyclerView.ViewHolder(view) {
        private val binding = ItemPeerBinding.bind(view)

        init {
            itemView.setOnClickListener {
                val position = bindingAdapterPosition
                if (position == RecyclerView.NO_POSITION) {
                    return@setOnClickListener
                }
                listener(getItem(position).peerId)
            }
        }

        fun bind(item: PeerUI) {
            binding.imgPeerAvatar.load(item.avatarfull) {
                transformations(CircleCropTransformation())
            }
            binding.tvPeerName.text = item.personaname
            binding.tvPeerGamesWith.text = item.withGames.toString()
            val winRate = (item.withWin.toDouble() / item.withGames) * 100
            binding.tvPeerWinRate.text =
                itemView.context.resources.getString(R.string.peer_winrate, winRate)
        }
    }

    companion object {
        private val DIFF_CALLBACK = object : DiffUtil.ItemCallback<PeerUI>() {
            override fun areItemsTheSame(oldItem: PeerUI, newItem: PeerUI): Boolean {
                return oldItem.peerId == newItem.peerId
            }

            override fun areContentsTheSame(oldItem: PeerUI, newItem: PeerUI): Boolean {
                return oldItem == newItem
            }
        }
    }
}
